Turns tablecrafting into personal crafting

Razharas opened the pull request:
Turns tablecrafting into personal crafting, you dont need tables anymore
Apart from being datum and opening by clicking on UI button its all the same externally
Internally a lot of crafting procs are rewritten, lots of useless shit thrown out, lots of useful shit thrown in
:cl:
tweak: crafting can now be done everywhere, little button near the intents with hammer on it(its not T) opens the crafting menu, if you are in a locker or mech only things in your hands will be considered for crafting
/:cl:

:star2: Like, comment, upvote & subscribe :star2:

Razharas committed:
Converts tablecrafting to datums

Razharas committed:
Adds crafting.dm

Adds crafting.dm

Razharas committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into TCfix

Conflicts:
	code/game/objects/items/weapons/twohanded.dm
	code/modules/crafting/recipes.dm

Merge conflicts

Razharas committed:
Fix merge conflicts

Fix merge conflicts

Razharas committed:
Fixes dme

RemieRichards commented on code/game/atoms.dm:
```diff
@@ -66,7 +66,15 @@
+				L.unEquip(M)
+			M.loc = src
+		contents += A
 	return
```
trailing return, remove it please.

RemieRichards commented on code/game/machinery/machinery.dm:
```diff
@@ -323,7 +323,8 @@ Class Procs:
+/obj/machinery/CheckParts(list/content)
+	..()
 	RefreshParts()
 	return
```
trailing return, remove it please.

Razharas replied:
Didnt even touch this code its all autoreplace

coiax replied:
Remie got beaten up by trailing returns at school

coiax commented:
We minecraft now

nullbear commented:
@coiax It's more like terraria crafting.

Speaking of which, good that you made it /atom/movable. Means you can have stuff like 'workbench' as a tool for building something. If the workbench machine is nearby, you can use it for a recipe. XP

Razharas committed:
Fixes returns

Fixes returns

Razharas commented:
Yes you can use anything as tool, even some specific turf or TABLE

Zilenan91 commented:
So this adds a crafting tab like the IC and OOC tab right? Or is it a hug button

Razharas commented:
Its a button in the upper left like toggle like and such

Fox-McCloud commented:
:+1: 

but please add a way to toggle the button on or off---action button clutter is a real thing.

nullbear commented on code/game/objects/items/weapons/twohanded.dm:
```diff
@@ -336,7 +336,18 @@
+	update_mob()
+	explosion(src.loc,-1,1,3)
+	qdel(src)
+
```
this grenade was made redundant in my C4 PR, as C4 is now a grenade, and can be used in crafting as such. No need to re-add it.

Razharas replied:
I just overrode conflicts with what was in my branch eh

nullbear replied:
i assumed as much. You might have messed up the chembomb recipe again then as well.

nullbear commented on code/modules/crafting/craft.dm:
```diff
@@ -0,0 +1,304 @@
+/datum/personal_crafting/proc/get_environment(mob/user)
+	. = list()
+	. += range(2, user.loc)
+	. += user.contents
```
check user.hands, not user.contents. Otherwise people will be able to do stuff like craft brainburgers using their own brain.

Razharas replied:
Feature

Iamgoofball replied:
feature

Razharas commented:
Theres a "collapse buttons" button there, but at least one will stay there

nullbear commented on code/modules/crafting/craft.dm:
```diff
@@ -0,0 +1,304 @@
+/datum/personal_crafting/proc/get_surroundings(mob/user)
+	. = list()
+	for(var/obj/item/I in get_environment(user))
+		if(istype(I, /obj/item/stack))
```
Check for proximity to the player. Or else you will be able to access things through walls/windows.

Razharas replied:
If ones to add proximity checks it should be in the get_environment proc to cut out whole turfs instead of per-item checks of the same stuff

nullbear replied:
okay so when you run through the content list. Do two checks.

One: If loc == user.loc
Two: If A.Adjacent(user)

The first one will allow you to access things inside a locker, if you are also in the locker.
The second one will handle all the buggy window stuff.

If something doesnt pass BOTH of those checks, remove it from the content list.

nullbear commented on code/modules/crafting/recipes.dm:
```diff
@@ -30,17 +30,25 @@
+	parts = list(/obj/item/weapon/c4 = 1)
+	time = 40
+	category = CAT_WEAPON
 
```
remove the entire lance2 recipe.

Iamgoofball replied:
why

nullbear replied:
because it's redundant, since the first recipe covers C4 now. And the second one is trying to use a path that is inaccessible to players.

nullbear commented on code/modules/crafting/recipes.dm:
```diff
@@ -323,7 +332,7 @@
 	time = 30
```
this is fine. Surprised that it didnt get overwritten, when the lance recipe did.

coiax commented:
Also, human organs are NOT in user.contents, the organs are in null space.

nullbear commented on code/modules/food&drinks/recipes/tablecraft/recipes_burger.dm:
```diff
@@ -177,15 +177,15 @@
 	name = "Spell burger"
 	reqs = list(
 		/obj/item/clothing/head/wizard/fake,
 	)
```
should be wizard/fake = 1

This was a bug i fixed, but that was reverted due to other issues.

Razharas committed:
Only check hands and remove old paths

Only check hands and remove old paths

nullbear commented:
that's everything i caught.

nullbear commented on code/modules/crafting/craft.dm:
```diff
@@ -31,7 +31,7 @@
 	. += range(2, user.loc)
-	. += user.contents
+	. += user.hands
 
```
try "get_both_hands(user)". It returns a list of the objects in both of the users hands.

Razharas committed:
Recipe fixes

Recipe fixes

Core0verload commented:
Add a way to start selfcrafting that does not involve action buttons. Maybe dragging self on self?

PKPenguin321 commented:
i think this would actually be an acceptable case to use a verb (or have a completely new UI element that's not an action button), but regardless, this is a really great PR and eliminates one of the biggest parts of https://github.com/tgstation/-tg-station/issues/16107
super fine job here raz, :tada:s for everybody

Razharas commented:
Didnt you guys whine that dragging icons is shit?

Lumipharon commented:
being able to collapse the action buttons is not satisfactory.
If I have a bunch of implants/powers (which DOES happen frequently for certain antags), then I don't want another button clogging it up that I'm not going to be using, and I certainly don't want to hide ALL the buttons.

RandomMarine commented:
Something that every spaceman ALWAYS has the ability to should really just be a verb, just like rest, sleep, etc. We don't have an action button for toggling suit sensors, why? Because it'll cluttering the action buttons 90% of the time.

jughu commented:
Just make it a verb like Me like *craft

Razharas commented:
Ok i can make a verb

TechnoAlchemisto commented:
I'd prefer it be a verb than be an icon, it's not like it'll need to be done quickly and conveniently. Shit, crafting shouldn't really be convenient, I think the requisite for a table was a good balancer for not being able to craft bombspears on-the-go

Razharas commented:
Actually no people are saying verbs are the devil

PKPenguin321 commented:
then make a dedicated UI button (as in NOT an action button, but still on the HuD)
also this really needs a changelog

RandomMarine commented:
Le verbs are bad is practically a meme from 2014-2015 when shitloads of things had object verbs that cluttered the tab.

This isn't tied to having any objects.
It's an ability any spaceman has at every moment.
It should be a verb.

RemieRichards commented:
@RandomMarine No verbs buddy, sorry. The current plan is to move it to a dedicated button on the HUD, not as an action button.

PKPenguin321 commented:
while i'm not opposed to a new UI button randommarine has a really good point

Core0verload commented:
Yes, it needs to be a small HUD button.

Bawhoppen commented:
I don't care if its a verb or a HUD button, but it should not under any circumstances be an action button

Hatfish commented:
Why not just put it in the IC tab or something?

RemieRichards commented:
@Hatfish that'd be a verb silly!

Bawhoppen commented:
verbs are good though to be fair

KazeEspada commented:
This should be a verb and a button. Since its highly innate, and not tied to any object.

Shadowlight213 commented:
:-1: if it becomes an action button or a really big hud icon. A verb would be ideal, but if it HAS to be a hud element, make it not an action button and relatively small.

phil235 commented on code/modules/crafting/craft.dm:
```diff
@@ -0,0 +1,304 @@
+
+/datum/personal_crafting/proc/get_environment(mob/user)
+	. = list()
+	. += range(2, user.loc)
```
using stuff through boder windows and windoors.

nullbear replied:
i posted a fix to this before already.

phil235 commented on code/modules/food&drinks/food/snacks.dm:
```diff
@@ -151,7 +152,10 @@
-/obj/item/weapon/reagent_containers/food/snacks/CheckParts()
+/obj/item/weapon/reagent_containers/food/snacks/CheckParts(list/content)
+	..()
+	create_reagents(volume)
```
snacks don't have create_reagents() in their New() ?

Razharas replied:
This was dont in the create_item proc after checking if the resulting shit a reagent container or something
I thought its needed

phil235 replied:
You need to put it before ..() Since the reagent transfer is done in reagent_containers/CheckParts()

phil235 commented on code/modules/reagents/reagent_containers.dm:
```diff
@@ -24,6 +24,13 @@
 
+/obj/item/weapon/reagent_containers/CheckParts(list/content)
+	..()
+	var/datum/reagents/RS = new(volume)
```
same thing, reagent_containers/New() doesn't handle reagents and volume creation?

Razharas replied:
As with snacks, there was special type check in the crafting code that did exactly this
I just moved the code to an appropriate place, dunno why its needed in the first place

Razharas committed:
Removed action button

Removed action button
Added small ui button to inventory near the intent wheel
Removed debug
href_list is not longer /list

Razharas committed:
Added adjucent checks

Added adjucent checks

phil235 commented on code/modules/reagents/reagent_containers.dm:
```diff
@@ -24,6 +24,13 @@
+/obj/item/weapon/reagent_containers/CheckParts(list/content)
+	..()
+	var/datum/reagents/RS = new(volume)
+	for(var/datum/reagent/R in contents)
```
Don't you mean "content" like the argument?

Razharas replied:
No, all the things in list/content are getting transfered into contents on on the atom level on checkparts

Razharas committed:
Snacks now get chems from parts

Snacks now get chems from parts

phil235 commented on code/modules/food&drinks/recipes/food_mixtures.dm:
```diff
@@ -1,3 +1,5 @@
+/datum/crafting_recipe/food/New()
+	parts = reqs
```
some food recipes starts with parts, so you have to do parts += reqs

phil235 commented on code/modules/food&drinks/food/snacks.dm:
```diff
@@ -151,7 +152,12 @@
+	..()
+	reagents.reagent_list.Cut()
+	for(var/obj/item/weapon/reagent_containers/RC in contents)
+		RC.reagents.trans_to(reagents, RC.reagents.maximum_volume)
```
You also need to make sure that items that were in reqs, but not parts originally, do not end up inside the result object.

Razharas replied:
Only parts end up in the content list
Reqs are deleted in the del_reqs proc

Razharas committed:
Complex food fuckery fixed

Complex food fuckery fixed

Razharas committed:
Fixes merge conflicts

Fixes merge conflicts

Razharas committed:
Fixes old recipe

Fixes old recipe

Incomptinence commented:
mite b gud

MrStonedOne commented:
Test merged, almost broke the server.

get_reagents was getting called ~80 million of times a second.

something fucked up. it seems to be related to spreading out ingredients across multiple tables

Razharas committed:
Fix to infinite loops in reagent extraction

Fix to infinite loops in reagent extraction
It was finding old empty containers and tries to extract from them

Razharas committed:
Fixes merge conflicts

Fixes merge conflicts

phil235 commented:
>Test merged, almost broke the server.
get_reagents was getting called ~80 million of times a second.

@Razharas You didn't even test it on a local server?

Razharas commented:
It was only happening when recipes had reagents reqs and the reagents were in the separate containers and each container had less than needed
Cases this specific i didnt test

Incomptinence commented:
I retract my thumbs up and party popper in light of this malfunction. 
Sorry Raz.

Razharas committed:
Fixes hands not checking

Fixes hands not checking

Razharas committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into TCfix

phil235 commented:
>if you are in a locker or mech only things in your hands will be considered for crafting

The things you craft will drop inside the mech and be lost forever. It might be more cautious to prevent crafting while not on a turf.

phil235 commented on code/modules/crafting/craft.dm:
```diff
@@ -0,0 +1,300 @@
+	return
+
+/datum/personal_crafting/Topic(href, href_list)
+	if(usr.stat || usr.lying)
```
incapacitated() instead of stat maybe.

phil235 commented:
I still managed to make the server nearly crash by crafting some crossbuns. Probably the same issue that mso mentionned.

Razharas committed:
Fixes reagents fucking up

Fixes reagents fucking up

Razharas committed:
Even more reagent fuck ups fixed and comments

Even more reagent fuck ups fixed and comments

Razharas committed:
Fixes all the shit i could find

Fixes all the shit i could find

Razharas committed:
Removes debug and mechs

Removes debug and mechs

phil235 commented on code/modules/projectiles/gun.dm:
```diff
@@ -76,7 +76,8 @@
 
 
-/obj/item/weapon/gun/CheckParts()
+/obj/item/weapon/gun/CheckParts(list/contents)
```
typo in the argument, "content" not "contents". This wouldn't happen if the arg name was not so similar to a var name.

Razharas committed:
Argument renaming

Argument renaming

phil235 commented on code/modules/projectiles/gun.dm:
```diff
@@ -76,7 +76,8 @@
 
 
-/obj/item/weapon/gun/CheckParts()
+/obj/item/weapon/gun/CheckParts(list/parts)
```
err, all CheckParts() should probably use the same names for their argument.

Razharas committed:
Fixes all argument names

Fixes all argument names

phil235 commented on code/modules/food&drinks/food/snacks.dm:
```diff
@@ -151,7 +152,18 @@
 
 //Called when you finish tablecrafting a snack.
-/obj/item/weapon/reagent_containers/food/snacks/CheckParts()
+/obj/item/weapon/reagent_containers/food/snacks/CheckParts(list/content, datum/crafting_recipe/R)
```
here too

phil235 commented on code/game/mecha/combat/combat.dm:
```diff
@@ -4,6 +4,7 @@
 	damage_absorption = list("brute"=0.7,"fire"=1,"bullet"=0.7,"laser"=0.85,"energy"=1,"bomb"=0.8)
 
 /obj/mecha/combat/CheckParts(atom/holder)
```
why is mech using the same proc ??

Razharas replied:
Checkparts is supposed to be univerlas proc for everything to check its contents, get useful shit form there and incorporate inside for some purpose
Its not hardcoded to be crafting-only, mech building and machine building can use it too

Razharas committed:
Final fix of arguments

Final fix of arguments
